apm-agent-plugins/apm-es-restclient-plugin/src/main/java/co/elastic/apm/es/restclient/ElasticsearchRestClientInstrumentation.java

```diff
@@ -0,0 +1,147 @@
+    static final String ELASTICSEARCH_NODE_URL_KEY = "Elasticsearch-node-url";
+    @VisibleForAdvice
+    static final String QUERY_STATUS_CODE_KEY = "Query-status-code";
+    @VisibleForAdvice
+    static final String ERROR_REASON_KEY = "Error-reason";
+
+    @Advice.OnMethodEnter
+    private static void onBeforeExecute(@Advice.Argument(0) Request request,
+                                        @Advice.Local("span") Span span) {
+        if (tracer == null || tracer.getActive() == null || !tracer.getActive().isSampled()) {
+            return;
+        }
+        span = tracer.getActive().createSpan()
+            .withType(SPAN_TYPE)
+            .appendToName("Elasticsearch: ").appendToName(request.getMethod()).appendToName(" ").appendToName(request.getEndpoint());
+        span.getContext().getDb().withType(DB_CONTEXT_TYPE);
+        span.activate();
+
+        if (span.isSampled()) {
+            span.getContext().getHttp().withMethod(request.getMethod());
+            if (request.getEndpoint().endsWith(SEARCH_QUERY_PATH_SUFFIX)) {
+                HttpEntity entity = request.getEntity();
+                if (entity != null && entity.isRepeatable()) {
+                    try {
+                        String body = ESRestClientInstrumentationHelper.readRequestBody(entity.getContent(), request.getEndpoint());
+                        if (body != null && !body.isEmpty()) {
+                            span.getContext().getDb().withStatement(body);
+                        }
+                    } catch (IOException e) {
+                        // We can't log from here
+                    }
+                }
+            }
+        }
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class)
+    public static void onAfterExecute(@Advice.Return @Nullable Response response,
+                                      @Advice.Local("span") @Nullable Span span,
+                                      @Advice.Thrown @Nullable Throwable t) {
+        if (span != null) {
+            try {
+                String url = null;
+                int statusCode = -1;
+                if(response != null) {
+                    url = response.getHost().toURI().toString();
+                    statusCode = response.getStatusLine().getStatusCode();
+                } else if(t != null) {
+                    if (t instanceof ResponseException) {
+                        ResponseException esre = (ResponseException) t;
+                        url = esre.getResponse().getHost().toURI().toString();
+                        statusCode = esre.getResponse().getStatusLine().getStatusCode();
+
+                        // Add tags so that they will be copied to error capture
+                        span.addTag(QUERY_STATUS_CODE_KEY, Integer.toString(statusCode));
+                        span.addTag(ELASTICSEARCH_NODE_URL_KEY, url);
+                        span.addTag(ERROR_REASON_KEY, esre.getResponse().getStatusLine().getReasonPhrase());
```

Comment:
Does that yield valuable information which is not in the exception message? What are examples of the reason phrase vs the exception message?